Fix Windows error in distributed

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60167 Fix Windows error in distributed**

We were getting errors such as this on Windows in our c10d ProcessGroup test suite:
```
  test_send_recv_all_to_all (__main__.ProcessGroupGlooTest) ... Exception in thread Thread-1:
Traceback (most recent call last):
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_distributed.py", line 471, in _event_listener
    if pipe.poll(None):
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 257, in poll
    return self._poll(timeout)
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 330, in _poll
    return bool(wait([self], timeout))
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 883, in wait
    ov.cancel()
OSError: [WinError 6] The handle is invalid
Fatal Python error: could not acquire lock for <_io.BufferedWriter name='<stderr>'> at interpreter shutdown, possibly due to daemon threads
Python runtime state: finalizing (tstate=000001EFDF228CE0)

Thread 0x00001f68 (most recent call first):
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 1202 in invoke_excepthook
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 934 in _bootstrap_inner
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 890 in _bootstrap

Current thread 0x00000f94 (most recent call first):
<no Python frame>
FAIL (5.009s)
```
And the process would then exit with error code 3221226505.
See: https://app.circleci.com/pipelines/github/pytorch/pytorch/337351/workflows/ad919a3e-fe9a-4566-8ad6-8b0a252f730c/jobs/14170191/steps

By looking at [the code of `_event_listener` in `common_distributed.py`](https://github.com/pytorch/pytorch/blob/eb36f67dcc7000caa58fa4dfa9c089ce17c6d523/torch/testing/_internal/common_distributed.py#L467-L489) I think that the first exception (the one about the handle being invalid) is "expected" as it results from another thread purposely closing the pipe while that thread is polling it.

The relevant part of the problem seems to be the "could not acquire lock" one. I think this stems from the event listener thread being launched as a daemon thread, which means the interpreter will not wait for that thread to complete before shutting down. When the interpreter shuts down it instantly aborts all other threads. If the event listener thread was aborter _while_ it was logging to stdout then that thread was holding the lock but never got to release it. This is probably what the error is complaining about. This seems to be intended/expected behavior for CPython: https://bugs.python.org/issue42717.

The solution thus is simple: don't make that thread a daemon thread and explicitly wait for it to terminate before shutting down.

Differential Revision: [D29193014](https://our.internmc.facebook.com/intern/diff/D29193014/)

lw committed:
Fix Windows error in distributed

We were getting errors such as this on Windows in our c10d ProcessGroup test suite:
```
  test_send_recv_all_to_all (__main__.ProcessGroupGlooTest) ... Exception in thread Thread-1:
Traceback (most recent call last):
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_distributed.py", line 471, in _event_listener
    if pipe.poll(None):
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 257, in poll
    return self._poll(timeout)
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 330, in _poll
    return bool(wait([self], timeout))
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 883, in wait
    ov.cancel()
OSError: [WinError 6] The handle is invalid
Fatal Python error: could not acquire lock for <_io.BufferedWriter name='<stderr>'> at interpreter shutdown, possibly due to daemon threads
Python runtime state: finalizing (tstate=000001EFDF228CE0)

Thread 0x00001f68 (most recent call first):
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 1202 in invoke_excepthook
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 934 in _bootstrap_inner
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 890 in _bootstrap

Current thread 0x00000f94 (most recent call first):
<no Python frame>
FAIL (5.009s)
```
And the process would then exit with error code 3221226505.
See: https://app.circleci.com/pipelines/github/pytorch/pytorch/337351/workflows/ad919a3e-fe9a-4566-8ad6-8b0a252f730c/jobs/14170191/steps

By looking at [the code of `_event_listener` in `common_distributed.py`](https://github.com/pytorch/pytorch/blob/eb36f67dcc7000caa58fa4dfa9c089ce17c6d523/torch/testing/_internal/common_distributed.py#L467-L489) I think that the first exception (the one about the handle being invalid) is "expected" as it results from another thread purposely closing the pipe while that thread is polling it.

The relevant part of the problem seems to be the "could not acquire lock" one. I think this stems from the event listener thread being launched as a daemon thread, which means the interpreter will not wait for that thread to complete before shutting down. When the interpreter shuts down it instantly aborts all other threads. If the event listener thread was aborter _while_ it was logging to stdout then that thread was holding the lock but never got to release it. This is probably what the error is complaining about. This seems to be intended/expected behavior for CPython: https://bugs.python.org/issue42717.

The solution thus is simple: don't make that thread a daemon thread and explicitly wait for it to terminate before shutting down.

Differential Revision: [D29193014](https://our.internmc.facebook.com/intern/diff/D29193014/)

[ghstack-poisoned]

lw committed:
Update on "Fix Windows error in distributed"


We were getting errors such as this on Windows in our c10d ProcessGroup test suite:
```
  test_send_recv_all_to_all (__main__.ProcessGroupGlooTest) ... Exception in thread Thread-1:
Traceback (most recent call last):
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_distributed.py", line 471, in _event_listener
    if pipe.poll(None):
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 257, in poll
    return self._poll(timeout)
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 330, in _poll
    return bool(wait([self], timeout))
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 883, in wait
    ov.cancel()
OSError: [WinError 6] The handle is invalid
Fatal Python error: could not acquire lock for <_io.BufferedWriter name='<stderr>'> at interpreter shutdown, possibly due to daemon threads
Python runtime state: finalizing (tstate=000001EFDF228CE0)

Thread 0x00001f68 (most recent call first):
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 1202 in invoke_excepthook
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 934 in _bootstrap_inner
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 890 in _bootstrap

Current thread 0x00000f94 (most recent call first):
<no Python frame>
FAIL (5.009s)
```
And the process would then exit with error code 3221226505.
See: https://app.circleci.com/pipelines/github/pytorch/pytorch/337351/workflows/ad919a3e-fe9a-4566-8ad6-8b0a252f730c/jobs/14170191/steps

By looking at [the code of `_event_listener` in `common_distributed.py`](https://github.com/pytorch/pytorch/blob/eb36f67dcc7000caa58fa4dfa9c089ce17c6d523/torch/testing/_internal/common_distributed.py#L467-L489) I think that the first exception (the one about the handle being invalid) is "expected" as it results from another thread purposely closing the pipe while that thread is polling it.

The relevant part of the problem seems to be the "could not acquire lock" one. I think this stems from the event listener thread being launched as a daemon thread, which means the interpreter will not wait for that thread to complete before shutting down. When the interpreter shuts down it instantly aborts all other threads. If the event listener thread was aborter _while_ it was logging to stdout then that thread was holding the lock but never got to release it. This is probably what the error is complaining about. This seems to be intended/expected behavior for CPython: https://bugs.python.org/issue42717.

The solution thus is simple: don't make that thread a daemon thread and explicitly wait for it to terminate before shutting down.

Differential Revision: [D29193014](https://our.internmc.facebook.com/intern/diff/D29193014/)

[ghstack-poisoned]

lw committed:
Update on "Fix Windows error in distributed"


We were getting errors such as this on Windows in our c10d ProcessGroup test suite:
```
  test_send_recv_all_to_all (__main__.ProcessGroupGlooTest) ... Exception in thread Thread-1:
Traceback (most recent call last):
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_distributed.py", line 471, in _event_listener
    if pipe.poll(None):
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 257, in poll
    return self._poll(timeout)
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 330, in _poll
    return bool(wait([self], timeout))
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 883, in wait
    ov.cancel()
OSError: [WinError 6] The handle is invalid
Fatal Python error: could not acquire lock for <_io.BufferedWriter name='<stderr>'> at interpreter shutdown, possibly due to daemon threads
Python runtime state: finalizing (tstate=000001EFDF228CE0)

Thread 0x00001f68 (most recent call first):
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 1202 in invoke_excepthook
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 934 in _bootstrap_inner
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 890 in _bootstrap

Current thread 0x00000f94 (most recent call first):
<no Python frame>
FAIL (5.009s)
```
And the process would then exit with error code 3221226505.
See: https://app.circleci.com/pipelines/github/pytorch/pytorch/337351/workflows/ad919a3e-fe9a-4566-8ad6-8b0a252f730c/jobs/14170191/steps

By looking at [the code of `_event_listener` in `common_distributed.py`](https://github.com/pytorch/pytorch/blob/eb36f67dcc7000caa58fa4dfa9c089ce17c6d523/torch/testing/_internal/common_distributed.py#L467-L489) I think that the first exception (the one about the handle being invalid) is "expected" as it results from another thread purposely closing the pipe while that thread is polling it.

The relevant part of the problem seems to be the "could not acquire lock" one. I think this stems from the event listener thread being launched as a daemon thread, which means the interpreter will not wait for that thread to complete before shutting down. When the interpreter shuts down it instantly aborts all other threads. If the event listener thread was aborter _while_ it was logging to stdout then that thread was holding the lock but never got to release it. This is probably what the error is complaining about. This seems to be intended/expected behavior for CPython: https://bugs.python.org/issue42717.

The solution thus is simple: don't make that thread a daemon thread and explicitly wait for it to terminate before shutting down.

Differential Revision: [D29193014](https://our.internmc.facebook.com/intern/diff/D29193014/)

[ghstack-poisoned]

lw committed:
Update on "Fix Windows error in distributed"


We were getting errors such as this on Windows in our c10d ProcessGroup test suite:
```
  test_send_recv_all_to_all (__main__.ProcessGroupGlooTest) ... Exception in thread Thread-1:
Traceback (most recent call last):
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_distributed.py", line 471, in _event_listener
    if pipe.poll(None):
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 257, in poll
    return self._poll(timeout)
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 330, in _poll
    return bool(wait([self], timeout))
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 883, in wait
    ov.cancel()
OSError: [WinError 6] The handle is invalid
Fatal Python error: could not acquire lock for <_io.BufferedWriter name='<stderr>'> at interpreter shutdown, possibly due to daemon threads
Python runtime state: finalizing (tstate=000001EFDF228CE0)

Thread 0x00001f68 (most recent call first):
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 1202 in invoke_excepthook
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 934 in _bootstrap_inner
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 890 in _bootstrap

Current thread 0x00000f94 (most recent call first):
<no Python frame>
FAIL (5.009s)
```
And the process would then exit with error code 3221226505.
See: https://app.circleci.com/pipelines/github/pytorch/pytorch/337351/workflows/ad919a3e-fe9a-4566-8ad6-8b0a252f730c/jobs/14170191/steps

By looking at [the code of `_event_listener` in `common_distributed.py`](https://github.com/pytorch/pytorch/blob/eb36f67dcc7000caa58fa4dfa9c089ce17c6d523/torch/testing/_internal/common_distributed.py#L467-L489) I think that the first exception (the one about the handle being invalid) is "expected" as it results from another thread purposely closing the pipe while that thread is polling it.

The relevant part of the problem seems to be the "could not acquire lock" one. I think this stems from the event listener thread being launched as a daemon thread, which means the interpreter will not wait for that thread to complete before shutting down. When the interpreter shuts down it instantly aborts all other threads. If the event listener thread was aborter _while_ it was logging to stdout then that thread was holding the lock but never got to release it. This is probably what the error is complaining about. This seems to be intended/expected behavior for CPython: https://bugs.python.org/issue42717.

The solution thus is simple: don't make that thread a daemon thread and explicitly wait for it to terminate before shutting down.

Differential Revision: [D29193014](https://our.internmc.facebook.com/intern/diff/D29193014/)

[ghstack-poisoned]

lw committed:
Update on "Fix Windows error in distributed"


We were getting errors such as this on Windows in our c10d ProcessGroup test suite:
```
  test_send_recv_all_to_all (__main__.ProcessGroupGlooTest) ... Exception in thread Thread-1:
Traceback (most recent call last):
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_distributed.py", line 471, in _event_listener
    if pipe.poll(None):
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 257, in poll
    return self._poll(timeout)
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 330, in _poll
    return bool(wait([self], timeout))
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 883, in wait
    ov.cancel()
OSError: [WinError 6] The handle is invalid
Fatal Python error: could not acquire lock for <_io.BufferedWriter name='<stderr>'> at interpreter shutdown, possibly due to daemon threads
Python runtime state: finalizing (tstate=000001EFDF228CE0)

Thread 0x00001f68 (most recent call first):
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 1202 in invoke_excepthook
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 934 in _bootstrap_inner
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 890 in _bootstrap

Current thread 0x00000f94 (most recent call first):
<no Python frame>
FAIL (5.009s)
```
And the process would then exit with error code 3221226505.
See: https://app.circleci.com/pipelines/github/pytorch/pytorch/337351/workflows/ad919a3e-fe9a-4566-8ad6-8b0a252f730c/jobs/14170191/steps

By looking at [the code of `_event_listener` in `common_distributed.py`](https://github.com/pytorch/pytorch/blob/eb36f67dcc7000caa58fa4dfa9c089ce17c6d523/torch/testing/_internal/common_distributed.py#L467-L489) I think that the first exception (the one about the handle being invalid) is "expected" as it results from another thread purposely closing the pipe while that thread is polling it.

The relevant part of the problem seems to be the "could not acquire lock" one. I think this stems from the event listener thread being launched as a daemon thread, which means the interpreter will not wait for that thread to complete before shutting down. When the interpreter shuts down it instantly aborts all other threads. If the event listener thread was aborter _while_ it was logging to stdout then that thread was holding the lock but never got to release it. This is probably what the error is complaining about. This seems to be intended/expected behavior for CPython: https://bugs.python.org/issue42717.

The solution thus is simple: don't make that thread a daemon thread and explicitly wait for it to terminate before shutting down.

Differential Revision: [D29193014](https://our.internmc.facebook.com/intern/diff/D29193014/)

[ghstack-poisoned]

lw committed:
Update on "Fix Windows error in distributed"


We were getting errors such as this on Windows in our c10d ProcessGroup test suite:
```
  test_send_recv_all_to_all (__main__.ProcessGroupGlooTest) ... Exception in thread Thread-1:
Traceback (most recent call last):
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_distributed.py", line 471, in _event_listener
    if pipe.poll(None):
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 257, in poll
    return self._poll(timeout)
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 330, in _poll
    return bool(wait([self], timeout))
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 883, in wait
    ov.cancel()
OSError: [WinError 6] The handle is invalid
Fatal Python error: could not acquire lock for <_io.BufferedWriter name='<stderr>'> at interpreter shutdown, possibly due to daemon threads
Python runtime state: finalizing (tstate=000001EFDF228CE0)

Thread 0x00001f68 (most recent call first):
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 1202 in invoke_excepthook
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 934 in _bootstrap_inner
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 890 in _bootstrap

Current thread 0x00000f94 (most recent call first):
<no Python frame>
FAIL (5.009s)
```
And the process would then exit with error code 3221226505.
See: https://app.circleci.com/pipelines/github/pytorch/pytorch/337351/workflows/ad919a3e-fe9a-4566-8ad6-8b0a252f730c/jobs/14170191/steps

By looking at [the code of `_event_listener` in `common_distributed.py`](https://github.com/pytorch/pytorch/blob/eb36f67dcc7000caa58fa4dfa9c089ce17c6d523/torch/testing/_internal/common_distributed.py#L467-L489) I think that the first exception (the one about the handle being invalid) is "expected" as it results from another thread purposely closing the pipe while that thread is polling it.

The relevant part of the problem seems to be the "could not acquire lock" one. I think this stems from the event listener thread being launched as a daemon thread, which means the interpreter will not wait for that thread to complete before shutting down. When the interpreter shuts down it instantly aborts all other threads. If the event listener thread was aborter _while_ it was logging to stdout then that thread was holding the lock but never got to release it. This is probably what the error is complaining about. This seems to be intended/expected behavior for CPython: https://bugs.python.org/issue42717.

The solution thus is simple: don't make that thread a daemon thread and explicitly wait for it to terminate before shutting down.

Differential Revision: [D29193014](https://our.internmc.facebook.com/intern/diff/D29193014/)

[ghstack-poisoned]

pritamdamania87 commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -465,16 +466,18 @@ class Event(Enum):
         logger.info(f'Starting event listener thread for {rank}')
         while True:
-            if pipe.poll(None):
+            ready_pipes = multiprocessing.connection.wait([parent_pipe, signal_pipe])
```
We probably need a timeout here since this is a non-daemon thread now? We don't want to block here forever?

pritamdamania87 replied:
Could this also throw an exception how `pipe.poll` did earlier?

lw replied:
> We probably need a timeout here since this is a non-daemon thread now? We don't want to block here forever?

I'm not sure we need a timeout:
- In the "normal" case (no deadlock), the main thread will call `signal_pipe.send(None)` at shutdown, which will make `signal_pipe` readable, causing this poll to return and the thread then to terminate.
- In the "deadlock" case the parent process will anyways end up killing the subprocess ungracefully, and I don't think it matters whether the thread is a daemon or not.

> Could this also throw an exception how pipe.poll did earlier?

It might, though this is not part of the problem. Just like before, those exceptions would clog up the log but be harmless. They also wouldn't interfere with the thread winding down because, well, an uncaught exception kills the thread.

pritamdamania87 commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -515,14 +524,22 @@ def run_test(self, test_name: str, pipe) -> None:
             getattr(self, test_name)()
             # Close pipe after done with test.
-            pipe.close()
+            if signal_pipe is not None:
+                signal_pipe.send(None)
+            if event_listener_thread is not None:
+                event_listener_thread.join()
+            parent_pipe.close()
```
Might be cleaner to put all the cleanup logic in a `finally` clause?

lw replied:
How do you envision that working? We're doing slightly different things in the two cases (in particular in the except case we're sending the traceback over the `parent_pipe` and exiting). I don't see a clean way to reconcile them into a single codepath.

pritamdamania87 replied:
I mean the signal_pipe and event_listener_thread conditions are the same in both cases right? That could go in a finally clause.

lw replied:
Ok, I thought that `sys.exit` would immediately kill the process (and thus had to go last), while actually it's "just" raising an exception. Hence I guess we can do further cleanup after exiting (which is weird but ok).

BTW do you know why we're sending the traceback to the parent in case of an exception, instead of just logging it? It seems a more convoluted way of getting the same information in the end.

pritamdamania87 commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -515,14 +524,22 @@ def run_test(self, test_name: str, pipe) -> None:
             getattr(self, test_name)()
             # Close pipe after done with test.
-            pipe.close()
+            if signal_pipe is not None:
```
In which cases would signal_pipe not be None? Shouldn't we update `distributed_test.py` to have this logic as well since it calls `run_test`?

lw replied:
Well, it is _not_ None when we call it from the `_run` method of this very class, which I modified just a few lines above.

However it _is_ None when called from here:
https://github.com/pytorch/pytorch/blob/700df82881786f3560826f194aa9e04beeaa3fd8/torch/testing/_internal/distributed/distributed_test.py#L531

Note that TestDistBackend._run is currently _not_ starting any event listener thread, hence the signal pipe isn't relevant in that scenario. It might make sense to add the event listener to TestDistBackend too (or, at that point, to some common superclass) but that's beyond the fix I'm carrying out here.

pritamdamania87 reviewed (commented):
Thanks for making this code more robust!

pritamdamania87 reviewed (approved):
LGTM, could you fix the lint issues before landing?

lw committed:
Update on "Fix Windows error in distributed"


We were getting errors such as this on Windows in our c10d ProcessGroup test suite:
```
  test_send_recv_all_to_all (__main__.ProcessGroupGlooTest) ... Exception in thread Thread-1:
Traceback (most recent call last):
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_distributed.py", line 471, in _event_listener
    if pipe.poll(None):
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 257, in poll
    return self._poll(timeout)
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 330, in _poll
    return bool(wait([self], timeout))
  File "C:\Jenkins\Miniconda3\lib\multiprocessing\connection.py", line 883, in wait
    ov.cancel()
OSError: [WinError 6] The handle is invalid
Fatal Python error: could not acquire lock for <_io.BufferedWriter name='<stderr>'> at interpreter shutdown, possibly due to daemon threads
Python runtime state: finalizing (tstate=000001EFDF228CE0)

Thread 0x00001f68 (most recent call first):
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 1202 in invoke_excepthook
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 934 in _bootstrap_inner
  File "C:\Jenkins\Miniconda3\lib\threading.py", line 890 in _bootstrap

Current thread 0x00000f94 (most recent call first):
<no Python frame>
FAIL (5.009s)
```
And the process would then exit with error code 3221226505.
See: https://app.circleci.com/pipelines/github/pytorch/pytorch/337351/workflows/ad919a3e-fe9a-4566-8ad6-8b0a252f730c/jobs/14170191/steps

By looking at [the code of `_event_listener` in `common_distributed.py`](https://github.com/pytorch/pytorch/blob/eb36f67dcc7000caa58fa4dfa9c089ce17c6d523/torch/testing/_internal/common_distributed.py#L467-L489) I think that the first exception (the one about the handle being invalid) is "expected" as it results from another thread purposely closing the pipe while that thread is polling it.

The relevant part of the problem seems to be the "could not acquire lock" one. I think this stems from the event listener thread being launched as a daemon thread, which means the interpreter will not wait for that thread to complete before shutting down. When the interpreter shuts down it instantly aborts all other threads. If the event listener thread was aborter _while_ it was logging to stdout then that thread was holding the lock but never got to release it. This is probably what the error is complaining about. This seems to be intended/expected behavior for CPython: https://bugs.python.org/issue42717.

The solution thus is simple: don't make that thread a daemon thread and explicitly wait for it to terminate before shutting down.

Differential Revision: [D29193014](https://our.internmc.facebook.com/intern/diff/D29193014/)

[ghstack-poisoned]
